packages/react/src/hooks/useReverification.ts

```diff
@@ -0,0 +1,146 @@
+import { reverificationMismatch } from '@clerk/shared/authorization-errors';
+import { ClerkRuntimeError, isClerkAPIResponseError, isClerkRuntimeError } from '@clerk/shared/error';
+import { useClerk } from '@clerk/shared/react';
+import type { __experimental_ReverificationConfig, Clerk } from '@clerk/types';
+import { useMemo } from 'react';
+
+interface PromiseWithResolvers<T> {
+  promise: Promise<T>;
+  resolve: (value: T | PromiseLike<T>) => void;
+  reject: (reason?: any) => void;
+}
+
+async function resolveResult<T>(result: Promise<T>): Promise<T | ReturnType<typeof reverificationMismatch>> {
+  return result
+    .then(r => {
+      if (r instanceof Response) {
+        return r.json();
+      }
+      return r;
+    })
+    .catch(e => {
+      // Treat fapi assurance as an assurance hint
+      if (isClerkAPIResponseError(e) && e.errors.find(({ code }) => code == 'session_step_up_verification_required')) {
+        return reverificationMismatch({} as __experimental_ReverificationConfig);
+      }
+
+      // rethrow
+      throw e;
+    });
+}
+
+/**
+ * Polyfill for Promise.withResolvers()
+ */
+function customPromiseWithResolves() {
+  let resolve: PromiseWithResolvers<unknown>['resolve'];
+  let reject: PromiseWithResolvers<unknown>['reject'];
+
+  const promise = new Promise((res, rej) => {
+    resolve = res;
+    reject = rej;
+  });
+
+  return {
+    promise,
+    // @ts-ignore
+    resolve,
+    // @ts-ignore
+    reject,
+  };
+}
+
+// Utility type to exclude anything that contains "clerk_error"
+type ExcludeClerkError<T> = T extends { clerk_error: any } ? never : T;
+
+const isReverificationHint = (result: any): result is ReturnType<typeof reverificationMismatch> => {
+  return (
+    result &&
+    typeof result === 'object' &&
+    'clerk_error' in result &&
+    result.clerk_error?.type === 'forbidden' &&
+    result.clerk_error?.reason === 'reverification-mismatch'
+  );
+};
+
+type InferParameters<T> = T extends (...args: infer P) => any ? P : never;
+type InferReturnType<T> = T extends (...args: any[]) => infer R ? R : never;
+
+function createReverificationHandler(params: { onOpenModal: Clerk['__experimental_openUserVerification'] }) {
+  function assertReverification<H extends (...args: InferParameters<H>) => InferReturnType<H>>(
+    fetcher: H,
+  ): (...args: InferParameters<H>) => Promise<ExcludeClerkError<Awaited<InferReturnType<H>>>> {
```

Comment:
❓  

Is `...args` supposed to be the callback the user passed in?  If so, I think we can probably simplify the TS types here by changing `...args` to `callback` as that'd remove all `array` types from the inferred types.

Also, looking at the example in the PR description I understood that the `handleReverification` just needs to return the same type as the return of the callback it accepts as the argument, is that correct?

Reply:
Seems like for the types to be inferred correctly below `return async (...args)` I need to explicitly type them again as `(...args: Parameters<Fetcher>)`